### DISCUSSION TITLE
Takte und das Energieintegral

### DISCUSSION THREAD

**Discussion by RotorMotor42** [cyberthom42/vaillant-arotherm-plus#4] [General]
Grundsätzlich sollen weniger und dafür längere Takte effizienter und besser für die Lebensdauer der WP sein.
Daher möchten wir die Anzahl Takte reduzieren.

Grunsätzlich wird das Takten über das Energieintegral (EI) gesteuert.
Leider lassen sich die Limits nur für den ersten Takt einstellen.
Danach ist es fest zwischen 0 und -60 (es wurden auch schon werte von -66 und mehr beobachtet, was aber Überschwinger sein können).

Außerdem wurde bereits festgestellt, dass das EI nicht einmal pro Minute mit VL_ist - VL_soll berechnet, sondern alle 30 Sekunden. Daher wird ein Takt bereits halb so früh beendet, wie erwartet.

Die Idee war nun selbst ein "Erweitertes Energie Integral" zu berechnen indem die gewünschte Temperatur enstprechend angehoben und später wieder gesenkt wird.
Allerdings steigt trotz angehobener Temperatur das EI weiter.
Bei weiterer Anhebung steigt dann die Kompressordrehzahl.

Wir suchen nun nach Ideen die Takte zu verlängern.

**Comment by Apolyxon** [comment]
Ich verstehe das mal so, dass hier nicht die Einstellungen per "Bordmittel" besprochen werden sollen, sondern wirklich Fragen zum Skripting, richtig? Für die Besten "Bordeinstellungen" gegen Takten in der Übergangszeit dann vermutlich ein eigener Thread.

**Reply by RotorMotor42** [reply]
Grundsätzlich bin ich da offen, wenn es eine Lösung ohne Eingriffe über den Bus gibt, würde ich die natürlich gerne nehmen. Getrennte Threads können wir aber gerne machen.

**Comment by cyberthom42** [comment]
Sorry, ich bin mit meinen 2 Kleinkindern momentan gut eingespannt, kann also immer etwas dauern :-)
Zur Vermeidung des Takten, also dass das Energieintegral bis auf 0 ansteigt, kann man die min VL Temp anheben. Ich hatte dazu ein Script geschrieben. Das Script funktioniert eigentlich soweit ganz gut, ist aber eher ein erster Ansatz. Können wir hier zusammen verbessern. Ich habe als einen Parameter die minimale Taktdauer definiert, damit lege ich fest, wie lange der Takt mindestens dauern soll. 
Was jetzt noch fehlt wäre auch für die Pause dann eine minimale Pause zu definieren. Dazu hatte ich mir schon Gedanken gemacht. Ich denke das einfachste wäre hier mit der Max VL Temp zu arbeiten. Das Integral fällt ja immer sehr schnell wieder ab, bei mir z.T. innerhalb von 15 Minuten. Das würde ich gern verlängern, sagen wir auf 1h. Also dann nach dem Ende des Taktes die Max VL Temp auch auf 25 oder 26 setzen. Das sollte dann passen. Zur Not Min VL und Max VL noch weiter runter. Man könnte auch mit dem Heizmodus arbeiten, aber konsequenter weise würde ich hier auch mit Max VL arbeiten.

**Reply by RotorMotor42** [reply]
Kein Problem und kein Stress in der Weihnachtszeit. ;-)

Für mich ist das Grundsätzliche Problem beim Setzen von Min und Max Vorlauf, dass ich diese Werte dann nicht tracken kann.
Sobald ich den Wert einmal überschrieben habe, weiß man nicht mehr was die Regelung aktuell vorgeben würde.
Daher kann ich den Wert dann nichtmehr selber aufintegrieren.

Daher hatte ich angefangen ein Script zu schreien, dass direkt auf z1DayTemp regelt.
Hier weiß ich, dass es normal auf 21Grad gesetzt ist und kann jede Änderung tracken.

Bei meinem Script war nur das Problem, dass das EI trotz Anhebung weiter steigt:
![vaillant-arotherm-plus-vwl-35-55-75-610793-1](https://user-images.githubusercontent.com/121292653/209350424-e38cb379-67d8-41c4-9486-7bf49b31eedf.png)

Das hatte mir jetzt erstmal die Lust verdorben. ;-)

**Reply by cyberthom42** [reply]
Ah, guter Punkt. Dann kann man hier aber mit dem Parameter Hc1ExcessTemp arbeiten, dieser verändert den Wert für die SOll VL nicht. Der Regeleffekt ist eigentlich der gleiche. Man muss hier die geeignete Änderungsmenge suchen. Ich hatte das auch mal soweit, dass es recht gut lief.

**Reply by cyberthom42** [reply]
Wenn wir dann schon einmal so tief in der Materie sind, könnte man versuchen gleich manuell die VL Temp zu setzen. Die Formeln kennen wir ja nun. Wir müssten nur noch eine Lösung für den Mittelwert finden, wenn die Soll VL unter 28 oder tiefer ist. Hier muss man dann die Länge des Taktes und die Pause Berechnen oder Tag / Nachtabsenkung berechnen oder messen. Ich habe bei mir ja auch in jedem Raum Sensoren und an VL und RL für jeden Heizkreis. Es gibt ja diesen Heizmodus, in dem man manuell die VL Temp setzen kann. Da muss ich nochmal in die Anleitung schauen, ob das nur mit einem bestimmten Hydraulikshema geht.

**Reply by RotorMotor42** [reply]
`Hc1ExcessTemp` habe ich gerade mal getestet. Verhält sich beim Heizen recht ähnlich wie alle Anhebungen, die ich bisher getestet habe. Also IE kann trotzem steigen und bei zu viel Erhöhung dreht der Kompressor hoch.

`Hc1ExcessTemp` lässt sich nicht negativ setzen. Daher ist es nicht geeignet um damit den "Aus-Takt" zu regeln/verlängern.

Direkt den Vorlauf setzen könnte spannend sein. Wäre allerdings auch nicht so anders als die DayTemp anhand der Formel "rückwärts" zu setzen. Alles was direkt auf den Vorlauf oder sogar den Kompressor geht bedeutet halt recht viel Arbeit und auch Risiko falls das eigene Script ausfällt.

Also bei `Hc1ExcessTemp` bräuchten wir dann noch eine eigene Lösung für die Aus-Takte.

Wo siehst du die Probleme beim direkten setzen von `z1DayTemp`?

**Reply by cyberthom42** [reply]
z1DayTemp hatte ich bisher nicht verwendet, da es bisher gefühlt nicht so ganz berechenbar war wie stark der Anstieg der VL sein soll (dafür haben wir ja nun eine Formel). Die notwendige minimale VL kann ich ja immer aktuell ermitteln. Bei der Ermittlung der aktuell nötigen minimalen DayTemp hatte ich mich bisher schwer getan, zumal ich nicht weiß, ob auch "krumme" Zahlen verarbeitet werden. 
Aber du hast natürlich Recht, wenn wir mit z1DayTemp auch sauber arbeiten können, hätte dies den Vorteil, dass man nach wie vor die DesiredTemp sieht.

**Reply by RotorMotor42** [reply]
Die Frage ist noch, wie wir die Formel invertiert bekommen?
Also so, dass man eine Vorlauftemperatur rein steckt und eine DayTemp raus bekommt?
Invertierung ist mir bisher nicht geglückt.

**Reply by Apolyxon** [reply]
Was meinst du mit invertieren? Nach T_Außen lösen?

$VL_{soll} = T_{wunsch} + (2,8 * Heizkurve^{0,8})*(T_{wunsch}- T_{außen})^{0,75}$

**Reply by Apolyxon** [reply]
$T_{Außen} = \frac{1}{196} \left( 196 \cdot T_{Wunsch} - 5\sqrt[3]{5} \cdot 14^{2/3} \cdot \left( \frac{VL-T_{Wunsch}}{Heizkurve^{4/5}} \right)^{4/3} \right)$

oder

$T_{Außen} \approx 0.00510204 \left( 196 \cdot T_{Wunsch} - 49.6644 \left( \frac{VL-T_{Wunsch}}{Heizkurve^{4/5}} \right)^{4/3} \right)$

**Reply by cyberthom42** [reply]
Es geht darum, die Formel nach T_Wunsch (also `z1DayTemp`) aufzulösen. Das scheint nicht einfach möglich zu sein.

**Reply by martin-vi** [reply]
Vielleicht interessant: Ich hab gestern gesehen, dass es auf der Vaillant Website einen multiMATIC 700 Simulator gibt. Fand ich ganz interessant, denn dort kann man die Raumtemperatur und die Heizkurve einstellen und der Simulator passt die VLsoll an. 

Aus Interesse, habe ich mal geschaut, wie sie die Simulation implementiert haben. Siehe da, der Java Script Code ist nicht mal obfuscated.

![vaillant_vlsoll_math](https://github.com/user-attachments/assets/03537240-9c82-4293-8ff9-b0f4c41939ea)

Evtl. finden sich dort noch spannende andere Sachen :eyes:

**Comment by cyberthom42** [comment]
![grafik](https://user-images.githubusercontent.com/115342533/209344971-6ceabcf7-7473-4bf2-89b1-2bf9c07aecbb.png)

**Comment by cyberthom42** [comment]
![grafik](https://user-images.githubusercontent.com/115342533/209345066-8cc4b4d3-9e90-44ae-8b0c-be6aff6bf789.png)

**Comment by RotorMotor42** [comment]
Ich habe jetzt als schnelle Maßnahme einfach mal ein Script geschrieben, dass die Heizung nach einem Takt auf "aus" setzt und nach einer Temperaturabhängigen Pause wieder auf "auto".

Folgende Vorteile hat diese Variante:
- war sehr schnell runter geschrieben
- Aus-Takte lassen sich nun beliebig lang halten
- An-Takte werden länger dadurch, dass das IE nach aus bei -100 startet und auch noch weiter abfallen kann.
- An-Takte werden länger weil die Wärme vom vorherigen Takt in der Pause besser abgegeben werden konnte und dadurch die Temperatur nun weniger schnell überschießt
- Wasserpumpe läuft bei "aus" nicht im Gegensatz zur normalen Pause.

Nachteile:
- Taktlänge könnte noch länger sein (-100 IE limitiert hier nach wie vor)
- Pause schwer zu tunen

Hier kann man das Verhalten nach Aktivierung sehen:
![image](https://user-images.githubusercontent.com/121292653/209477825-01ea490a-29dc-4294-923a-2eeda3faeb41.png)

**Reply by cyberthom42** [reply]
Das ist auch eine sehr gute Idee. Heute war wieder ein guter Tag das zu testen. Ich hatte Vormittags mein Script noch aus, danach eingeschaltet. Das Ergebnis sieht dann so aus: (aktiviert um 10:15, 12:30 ist WW eingestellt)
![grafik](https://user-images.githubusercontent.com/115342533/209585377-14efb765-b1b4-4f27-95bf-f903ce4af758.png)
 
Ich denke die Challenge bei beiden Varianten ist, einen Weg zu finden 

1. auf den Tag gemittelt die erforderliche Wärmemenge zu finden. Ich vermute ein Lösungsweg wäre hier eine optimale RL zu bestimmen, nach dieser kann man ja immer tracken. Vielleicht macht es ja mal Sinn die Spreizung bei 5 Grad (bzw. die höchste Temp ohne Takten), bei 0 Grad, bei -10 Grad anzuschauen und dann nach oben zu interpolieren?
2. Sinnvolle Taktlängen / Pausen zu finden, in denen die Räume nicht auskühlen. Ich habe ja Sensoren im VL/RL und den Räumen. Ich wollte hierzu aber auch noch in einem Raum Sensoren auf den Fussboden anbringen. Dann kann ich den Abfall der Fussbodentemperatur beobachten. Die Temperatur wird ja früher abfallen als die Raumtemperatur.

**Comment by RotorMotor42** [comment]
Das Taktverlängern scheint ja ganz gut zu funktionieren. 
Führt aber ja schon zu einer Überhitzung.

Die genannten Punkte finde ich auch nicht einfach.

Ähnliche Probleme hatte man mit der Standardregelung aber auch schon:
1. Verluste durch Flüstermodus nachts
2. Verluste durch Warmwasser erzeugung
3. Gewinne durch solare Gewinne

Ich denke man muss eh in irgendeiner Form die Innentemperatur mit in die Regelung nehmen um eine konstante Innentemperatur zu erreichen.

**Reply by cyberthom42** [reply]
Steuerung nach Innentemperatur wird sicherlich zu träge, wobei man hier mal die Vaillant Steuerung also Raumaufschaltung testen könnte. Aber mitloggen auf jeden Fall. 
Mein aktueller pragmatischer Ansatz wäre folgender:

- Taktdauer definieren (evtl. abhängig von AT)
- Anhand der durchschnittlichen VL für den letzten Takt die theoretisch für diesen Takt einzustellende RT ermitteln
- Anhand der Länge des letzten Taktes und der errechneten RT dann die Taktpause errechnen um im Mittel auf gewünschte RT zu kommen. Im besten Fall wäre es ein langer Takt und eine lange Pause

**Comment by cyberthom42** [comment]
Ich habe heute endlich meinen PV-Ready Kontakt angeschlossen. Als dieser dann durch den Wechcselrichter bei Sonne geschlossen wurde, habe ich folgendes beobachtet: 
VL Temp steigt, aber `Hc1ActualFlowTempDesired` wird nicht angehoben. Es wird jedoch die Soll VL Temp im Feld `hmu SetMode` angehoben. Das Energieintegral fällt anfangs obwohl VL höher als `Hc1ActualFlowTempDesired` war. 
Meine Schlussfolgerung: Das Energieintegral wird mit dem Wert unter der Soll VL des Feldes `hmu SetMode` ermittelt und nicht mit dem Wert `Hc1ActualFlowTempDesired` . Ich hatte bisher kein Logging für die Soll VL im SetMode aktiviert, da mir der bisher immer zu ungenau war (0,5er Schritte). Aber das könnte einige Effekte erklären, die ich in der Vergangenheit gesehen habe (Energieintegral steigt, obwohl VL niedriger als VL Soll).

**Reply by RotorMotor42** [reply]
Das klingt natürlich spannend!
Wie zerlegt/loggt man denn überhaupt die ; getrennten Felder?
Gibt es das fertig oder muss ich das selber parsen/zerlegen?

**Reply by cyberthom42** [reply]
Ich mache alles komplett über meine Smarthome Steuerung in iobroker. Dort habe ich einen ebus Adapter, der trennt die Felder automatisch in einzelne Objekte. Ich hatte auch testweise mal ein eigenes Script gebastelt, das sah dann so aus:
![image](https://user-images.githubusercontent.com/115342533/209814730-f65e334e-d0c7-4640-acad-9ed21ab4ce01.png)

Als Javascript würde das dann so aussehen:

var Command, ListeResult, ListeState;
Command = ['read -f EnergyHcDay',',','read -f CurrentYieldPower',',','read -f HE_Vorlauftemperatur','','','','','',''].join('');
schedule("*/15 * * * * *", async function () {
  setState("ebus.0.cmd"/*ebusd command*/, Command);
  on({id: "ebus.0.cmdResult"/*ebusd command result*/, change: "any"}, async function (obj) {
    var value = obj.state.val;
    var oldValue = obj.oldState.val;
    ListeResult = (obj.state ? obj.state.val : "").split(',');
    console.warn(('ListeErstellen' + ''));
    setState("ebus.0.hmu.messages.EnergyHcDay.fields.0.value"/*data*/, parseFloat((ListeResult[0])), true);
    setState("ebus.0.hmu.messages.CurrentYieldPower.fields.0.value"/*data*/, parseFloat((ListeResult[1])), true);
    setState("ebus.0.hmu.messages.HE_Vorlauftemperatur.fields.0.value"/*data*/, parseFloat((ListeResult[2])), true);
  });
});

**Comment by martin-vi** [comment]
Kurzer Hinweis: Bei mir läuft seit rund einem Jahr eine Vaillant Arotherm mit dem Energie-Integral (WP geht an bei -60) Problem. Ich habe jetzt zwei neue Platinen bekommen (Garantie) für Außen und Innen-Einheit. 

```
# alt: vendor, circuit, HW rev, SW rev
Vaillant;HMU00;0517;5103
Vaillant;70000;0614;6903
Vaillant;VWZ00;0517;5103
# neu
Vaillant;HMU00;0902;5103
Vaillant;70000;0614;6903
Vaillant;VWZIO;0902;5103
```

Seit dem wird das Energie-Integral nicht mehr so schnell abgebaut und die WP fährt ordentlich bis Energie-Integral -100 (wie konfiguriert).
